### DISCUSSION TITLE
moment('24:00', 'HH:mm', true).isValid() returns true . which is not correct. There is no time 24:00, but 00:00 is correct

### DISCUSSION THREAD

**Discussion by ashwanikumar415** [moment/moment#6432] [Q&A] [answered]
**Describe the bug**
Even with latest moment library moment('24:00', 'HH:mm', true).isValid() returns true . There is no such time as 24:00 , its 00:00 
running below line in https://npm.runkit.com/moment  will show case behavior 
console.log(moment('24:00', 'HH:mm', true).isValid())

**To Reproduce**
1. Go to https://npm.runkit.com/moment
2. Run console.log(moment('24:00', 'HH:mm', true).isValid())
3. See value returning true

**Expected behavior**
Should return false, as there is no such time as 24:00
**version** 
2.27.0

Is there any alternate arguments/API that return correct results?

**Comment by maximillianfx** [comment]
Hi! I'm looking to an issue to contribute in this project (my first contribution 😄 ). Can I take this if this is really an issue?
PS.: I have executed this sample in my machine (node 12.18.2 and moment 2.27.0) and get the result `true` for `00:00` and `24:00`.

**Comment by maximillianfx** [comment]
So, @ashwanikumar415 , looking into the code, this time configuration is valid because the `moment` verify this especific value (24:00) and add 1 day in your datetime. So, in your example, your date is `now`(for example, today, 2020/07/22), so the moment will add 1 day in your date (2020/07/23) to accomplish the 24:00 time.

I don't know if this is really an issue, because these specific outlier is being covered in the source code.

**Comment by MinusFour** [comment]
It's part of the ISO-8601 spec `24:00` should be valid. However, according to Wikipedia, the new revision of the spec says it's no longer valid. Perhaps a compromise could be that it's not a valid date under strict parsing or maybe `moment` should somehow allow different rule sets for ISO-8601 versions.

**Comment by maximillianfx** [comment]
I think that the solution using the `strict` mode is more clean and unambiguous. Can I proceed with this change?

**Comment by maximillianfx** [comment]
I'm sorry, reading more about the ISO 8601-1 I could see that this change is critical and mandatory because it replaces the old revision. The solution could be an `invalidDate` when cheking if the `hour` field is equal to 24.

**Comment by maximillianfx** [comment]
Hi, in the `function checkOverflow()`, the clause to verify an overflow in the time is

```javascript
a[HOUR] < 0 || a[HOUR] > 24 || (a[HOUR] === 24 && (a[MINUTE] !== 0 || a[SECOND] !== 0 || a[MILLISECOND] !== 0))
```

If we change the `a[HOUR] > 24` to `a[HOUR] >= 24`, the problem can be solved. What you think?

**Comment by marwahaha** [comment]
Because of https://momentjs.com/docs/#/-project-status/ , I am not sure Moment.js should change behavior here.
cc @maggiepint

**Comment by mattjohnsonpint** [comment, marked as answer]
Thanks all for the thoughtful investigation.

Moment intentionally accepts exactly `24:00:00.000` as end-of-day and advances to the next day; values such as `24:01` remain invalid. If an application forbids `24:00`, I'd add an application-level check before parsing.

Given Moment's compatibility requirements, I don't think changing this established behavior is safe, so I'm closing it as answered.
